files/fail2ban-abuseipdb.sh

```diff
@@ -0,0 +1,255 @@
+#   - LOG_FILE: Path to the log file where actions and events are recorded.
+#
+# Dependencies:
+#   - curl: For making API requests to AbuseIPDB.
+#   - jq: For parsing JSON responses.
+#   - flock: Prevent data corruption.
+#
+# General Considerations:
+#   - The script does not interact with or rely on Fail2Ban’s SQLite database, as the Fail2ban database setup can vary across different environments.
+#   - The script performs two API calls to AbuseIPDB for each ban action in order to determine whether the IP should be reported.
+#   - First Call:  /v2/check → Checks if the IP is already reported.
+#   - Second Call: /v2/report → Reports the IP if necessary.
+#   - These two endpoints have separate daily limits, so does not affect your reporting quota.
+#   - When multiple instances of the script try to write to the REPORTED_IP_LIST_FILE concurrently we need to prevent data corruption. 'flock' used for that reason.
+#   - Never delete, truncate or try to sync REPORTED_IP_LIST_FILE with Fail2ban SQLite.
+#   - Even you reset fail2ban SQLite, continue to keep REPORTED_IP_LIST_FILE same, It serves as the script’s local tracking system.
+#   - When you go production keep watching '/var/log/abuseipdb-report.log' for any abnormal fails.
+#
+# Return Codes:
+#   0 - IP is reported
+#   1 - IP is not reported
+#
+# Exit Codes:
+#   1 - API-related failure
+#
+# Author:
+#   Hasan ÇALIŞIR - PSAUXIT
+#   hasan.calisir@psauxit.com
+#   https://www.psauxit.com/
+
+# Arguments
+APIKEY="$1"
+COMMENT="$2"
+IP="$3"
+CATEGORIES="$4"
+BANTIME="$5"
+
+# Configuration
+REPORTED_IP_LIST_FILE="/etc/fail2ban/action.d/abuseipdb-reported-ip-list"
+LOG_FILE="/var/log/abuseipdb-report.log"
+
+# Set defaults
+is_found_local=0
+shouldBanIP=1
+
+# Log messages
+log_message() {
+    local message="$1"
+    echo "$(date +"%Y-%m-%d %H:%M:%S") - ${message}" >> "${LOG_FILE}"
+}
+
+# Arguments validation
+if [[ -z "$1" || -z "$2" || -z "$3" || -z "$4" || -z "$5" ]]; then
+    log_message "FATAL: Usage: $0 <APIKEY> <COMMENT> <IP> <CATEGORIES> <BANTIME>"
+    exit 1
+fi
+
+# Ensure the directory for the reported IP list exists
+REPORTED_IP_DIR=$(dirname "${REPORTED_IP_LIST_FILE}")
+if [[ ! -d "${REPORTED_IP_DIR}" ]]; then
+    mkdir -p "${REPORTED_IP_DIR}" || { log_message "FATAL: Could not create directory ${REPORTED_IP_DIR}"; exit 1; }
+fi
+
+# Ensure the reported IP list file exists and is writable
+if [[ ! -f "${REPORTED_IP_LIST_FILE}" ]]; then
+    touch "${REPORTED_IP_LIST_FILE}" || { log_message "FATAL: Could not create file ${REPORTED_IP_LIST_FILE}"; exit 1; }
+fi
+
+if [[ ! -w "${REPORTED_IP_LIST_FILE}" ]]; then
+    log_message "FATAL: File ${REPORTED_IP_LIST_FILE} is not writable."
+    exit 1
+fi
+
+# Check runtime dependencies
+dependencies=("curl" "jq" "flock")
+for dep in "${dependencies[@]}"; do
+    if ! command -v "${dep}" &>/dev/null; then
+        log_message "FATAL: -${dep} is not installed. Please install -${dep} to proceed."
+        exit 1
+    fi
+done
```

Comment:
Firstly, I don't think it is good idea to check the dependencies by every ban (it'd be probably OK on actionstart, but definitelly not on actionban/actionunban).

Reply:
I agree. While `curl` and `flock` (provided by `util-linux`) are usually available even in slim images, `jq` often isn’t and  potentially cause issues, maybe we can keep it.

Reply:
I wasn't speaking about some particular dependency, the checks are basically OK...
What I wanted to tell you - it is worse to do that on every ban (in `actionban`), where it can be done once in `actionstart` (executed only at start of jail).
Since it is intended to be synchronous action (why actually not `nohup ... &`?) it'd delay banning action, and then every extra check is an additional reason for extra latency...
Just put it to `actionstart` and it'd be done only once.
However in my opinion the description about dependencies is fully enough.

Reply:
Sorry I don't have deep knowledge how things trigger in Fail2Ban but I think I got the idea and re-organized .sh. Please let me know If I missed something.